### PR TITLE
[inductor] Implement max_pool2d_with_indices as a reduction for large window sizes

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148210
* #148209
* __->__ #147876

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Comment by eellison** [comment]
perf dashboard run ?

**Comment by isuruf** [comment]
https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Wed%2C%2019%20Feb%202025%2013%3A14%3A00%20GMT&stopTime=Wed%2C%2026%20Feb%202025%2013%3A14%3A00%20GMT&granularity=hour&mode=inference&dtype=bfloat16&lDeviceName=cuda%20(a100)&rDeviceName=cuda%20(a100)&lBranch=main&lCommit=bb7e8fbd668c7c8931436b4a935b26911cbe0daf&rBranch=gh/isuruf/124/orig&rCommit=fe7db5cca7cce001b20d0fd9ccd5a4a1076e557f

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4321,20 +4321,16 @@ def max_pool2d_checks(
...
-@register_lowering(prims._low_memory_max_pool2d_with_offsets, type_promotion_kind=None)
-def _low_memory_max_pool2d_with_offsets(
+def _max_pool2d_with_offsets(
     x,
     kernel_size,
     stride,
     padding,
     dilation,
     ceil_mode=False,
+    assert_fallback=None,
```
Comment (eellison) [inline torch/_inductor/lowering.py:4331]: do we still need this arg ?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -1307,7 +1308,7 @@ def value_fn(
                 rindex = [sympy.expand(i) for i in rindex]
                 return (
                     inner_fn(index, rindex),
-                    ops.index_expr(flatten_index(rindex), torch.int64),
+                    ops.index_expr(flatten_index(rindex), dst_dtype),
```
Comment (eellison) [inline torch/_inductor/ir.py:1311]: Do we get the same perf if we cast the indices to int8 after ?

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -4354,44 +4350,74 @@ def _low_memory_max_pool2d_with_offsets(
...
+    kernel_size, stride, padding, dilation, _ = max_pool2d_checks(
+        x,
+        kernel_size,
+        stride,
+        padding,
+        dilation,
+        assert_fallback=False,
+    )
+
+    with config.patch(unroll_reductions_threshold=25):
```
Comment (eellison) [inline torch/_inductor/lowering.py:4410]: Nice way of deduping code !

File: torch/_inductor/lowering.py [hunk 4]
```diff
@@ -4429,8 +4455,40 @@ def offsets_to_indices(idx):
...
+    ceil_mode=False,
+):
+    kernel_size, stride, padding, dilation, _ = max_pool2d_checks(
+        x, kernel_size, stride, padding, dilation
+    )
+
+    if any(d > 1 for d in dilation):
+        return fallback_max_pool2d_with_indices(
+            x, kernel_size, stride, padding, dilation, ceil_mode=ceil_mode
+        )
```
Comment (eellison) [inline torch/_inductor/lowering.py:4481]: Would be cool follow up to handle dilation, so that we always lower max_pool2d..

**Review by eellison — COMMENTED** [review]
Very cool ! was just looking over a vision model with @exclamaforte the other day and seeing all the unlowered max_pool2d. i think the large window was probably the most common fallback we had among dashboard models, so its great to get these lowered.

If this pr is not based on a commit that is on the dashboard, would you mind doing a dashboard run of the base commit to reduce noise ?

**Comment by isuruf** [comment]
> If this pr is not based on a commit that is on the dashboard, would you mind doing a dashboard run of the base commit to reduce noise ?

The parent of the commit I used (fe7db5cca7) is bb7e8fbd668c7c8931436b4a935b26911cbe0daf and it is on the dashboard.

File: torch/_inductor/ir.py (hunk 2 above)
Comment (isuruf) [reply torch/_inductor/ir.py]: Yes, didn't see any perf change

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (isuruf) [reply torch/_inductor/lowering.py]: No, I removed it.

Comment (amjames) [reply torch/_inductor/lowering.py]: We shouldn't need it anymore. It is only consumer is the `max_pool2d_checks` fn, and the call to that has been removed.

**Review by eellison — COMMENTED** [review]
The perf dashboard is kind of odd.. big compilation time increase in cudagraphs but not default, & speedups in cudagraphs in HF training but not default 🤔 

Although the timm inference 2% is consistent and that's great !  

One of the models `DebertaV2ForMaskedLM ` shows a huge speedup but also a huge compilation time increase.. but only for cudagraphs not default 🤔  would you mind rebasing to 4216478250 (latest dashboard) and running again ? would be nice to have a bit more clarity before landing.

but looks great

**Comment by isuruf** [comment]
Here's a dashboard run with the direction correct.

https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Thu%2C%2020%20Feb%202025%2016%3A29%3A34%20GMT&stopTime=Thu%2C%2027%20Feb%202025%2016%3A29%3A34%20GMT&granularity=hour&mode=inference&dtype=bfloat16&deviceName=cuda%20(a100)&lBranch=gh/isuruf/124/orig&lCommit=fc4b07db4e7acbb1007289d7d8e1c28b169a2466&rBranch=main&rCommit=4216478250e08e950fdd090fc23a1b270c520cc4

**Comment by eellison** [comment]
On training [dm_nfnet_f0](https://hud.pytorch.org/benchmark/timm_models/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Thu,%2020%20Feb%202025%2016:29:34%20GMT&stopTime=Thu,%2027%20Feb%202025%2016:29:34%20GMT&granularity=hour&mode=training&model=dm_nfnet_f0&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/isuruf/124/orig&lCommit=fc4b07db4e7acbb1007289d7d8e1c28b169a2466&rBranch=main&rCommit=4216478250e08e950fdd090fc23a1b270c520cc4) and [timm_efficientnet](https://hud.pytorch.org/benchmark/torchbench/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Thu,%2020%20Feb%202025%2016:29:34%20GMT&stopTime=Thu,%2027%20Feb%202025%2016:29:34%20GMT&granularity=hour&mode=training&model=timm_efficientnet&dtype=amp&deviceName=cuda%20(a100)&lBranch=gh/isuruf/124/orig&lCommit=fc4b07db4e7acbb1007289d7d8e1c28b169a2466&rBranch=main&rCommit=4216478250e08e950fdd090fc23a1b270c520cc4) are showing no longer passing accuracy. would you mind verifying if this is real / bump tolerances if necessary?

**Comment by isuruf** [comment]
Both of those randomly fail for this PR and 4216478250e08e950fdd090fc23a1b270c520cc4

File: torch/_inductor/lowering.py [hunk 5]
```diff
@@ -4379,44 +4374,74 @@ def _low_memory_max_pool2d_with_offsets(
...
-        ):
-            ih = bh * stride[0] + h_inc - padding[0]
-            iw = bw * stride[1] + w_inc - padding[1]
-            val = x_loader([*prefix, ih, iw])
-            if return_index:
-                index = ops.index_expr(h_inc * kernel_size[1] + w_inc, torch.int8)
-                if maxindex is None:
-                    maxindex = index
-                else:
-                    maxindex = ops.where(ops.gt(val, maxval), index, maxindex)
```
Comment (isuruf) [inline torch/_inductor/lowering.py:4397]: This is slightly different in this PR as `argmax` does some nan checks between `val` and `maxval`

**Review by eellison — APPROVED** [review]
Very nice pr 🚀
